libbeat/reader/readjson/json.go

```diff
@@ -127,41 +141,47 @@ func createJSONError(message string) common.MapStr {
 	return common.MapStr{"message": message, "type": "json"}
 }
 
-func (pp *JSONPostProcessor) Name() string {
-	return "json"
-}
+// Next decodes JSON and returns the filled Line object.
+func (p *JSONParser) Next() (reader.Message, error) {
+	fieldsKey := "json" + p.id
```

Comment:
Glad you don't have to redo the json reader :), but I wonder if we should be a little more aggressive the the changes we want to add.

Potentially auto-generatiing the namespace is a huge issue. The namespace is not necessarily stable between restarts of the input. If we really need to auto generate it, we should add a sequence number for the levels. Like `json-1`, `json-2`. This allows users to discover the different layers, while keeping a stable way to query the data.

Right now we have the `fields_under_root: <bool>` setting. As you may have notived, I'm no fan of boolean in general... Booleans can be ambiguous, a set of boolean settings often don't compose well, we have to use pointers if we have 3 instead of 2 states (not configured, true, false), and they are not really extensible, potentially limiting features.

How about this. We make the `key` or `target` (the setting used in [LS json filter](https://www.elastic.co/guide/en/logstash/current/plugins-filters-json.html) and [Ingest Node] (https://www.elastic.co/guide/en/elasticsearch/reference/current/json-processor.html)) configurable. If `target` is empty, we merge the fields into the root document. Otherwise we use the configured keys.

Let's take and application that produces json logs and is run via docker for example. The integration might produce an input configuration like:

```
- type: filestream
  parsers:
    - ndjson:
        target: container
    - ndjson:
        target: application
        from: ... # optionally configure source field?
```

Now the event has 3 namespaces. The root fields from FB, the container namespace from the docker logs envelope, and the application namespace for the application logs.

Reply:
I named the source field `field` to be consistent with the other processors `fields` option that expect a list of field names.